spec/std/http/spec_helper.cr

```diff
@@ -2,7 +2,12 @@ require "spec"
 require "../spec_helper"
 require "../../support/fibers"
 
-private def wait_for(timeout = 5.seconds)
+private def wait_for
+  timeout = {% if flag?(:interpreted) %}
+              25.seconds
```

Comment:
What would you suggest? Let's say we go with 10. It times out. We try with 15. It times out. We keep bumping it. Is the timeout necessary at all in these specs if we expect them to pass?

Reply:
We expect specs to pass. But if they break for some reason, I wouldn't want to wait for the specs to finish for an accumulation of perhaps many minutes in total. For that reason the timeout should be chosen as small as possible while still ensuring specs to pass.

Reply:
Makes sense. I don't know what's a good value. Maybe we can change it to 10 seconds.

Reply:
On a related note, all specs have a hard-coded timeout of 15 seconds for the `-Dpreview_mt` workflow, and every now and then one spec fails. Example: https://github.com/crystal-lang/crystal/runs/7470644243?check_suite_focus=true

Reply:
I don't see a well-reasoned explanation from @straight-shoota that would be enough to oppose our two opinions.

We don't need to have tests that sometimes fail just to see that they are slow, there are much more direct ways.

> For that reason the timeout should be chosen as small as possible while still ensuring specs to pass.

That's just not the reason that people add timeouts to tests. It's so the test doesn't get stuck forever. The reason to lower the timeout is just if it *does* get stuck forever, you find out about that reasonably soon. But timeouts should still be set in a way that there's absolutely no chance that the test will sometimes be slow enough to reach the timeout if it's not actually stuck.

Reply:
> But timeouts should still be set in a way that there's absolutely no chance that the test will sometimes be slow enough to reach the timeout if it's not actually stuck.

That was exactly my thought process. I thought about doubling the time. But the spec should pass. So bumping the time a bit more it's fine. It won't take 25 for the spec to pass. If there's a problem with the spec, though, it will only take 25 seconds to find out, not more.

We could even bump it to 1 minute and it won't make current specs slower. We'll just have to wait 35 seconds more to find out if a spec is stuck.

Reply:
Ah actually I see that I misunderstood the message by @straight-shoota . It seems we agree on the main purpose of these timeouts, as he mentions "if they break". Then the difference is just that I suggested to err the other way.

Reply:
I'd argue that if a tested operation takes an incredibly long amount of time, I'd rather have it fail and be reported. This *might* just be a rare edge case caused by extreme unrelated circumstances. But it might also be an indicator for an error in the implementation that still makes it work eventually, but is unreasonably inefficient.

25 seconds are incredibly long for getting an HTTP server to start listening (which actually doesn't even involve any IO). It's just about getting the fiber running the server loop to start. That's shouldn't take 25 seconds, even in the interpreter. Even 5 seconds seems like way overkill for that (assuming specs don't run concurrently).

Reply:
Right, I'm sure the interpreter can do it in 5 seconds or much less... if it's complied in release mode. But in CI the interpreted is not optimized, and the non-optimized version is much, much slower than the optimized version (let's say, like 100 or 1000 times slower)

Reply:
I just ran some quick test and with a non-release compiler in interpreter mode, it takes about 22-24 ms for `server.listening?` to respond with `true`. It's basically just switching to the enqueued fiber, doing a syscall and switching back.
In compile mode, it's more like 0.1-0.2 ms. So it's (only) 100x faster than the non-release interpreter.
So if it happens to exceed the 5 seconds timeout in CI, it takes a lot longer than expected and might be worth looking into the reasons for that instead of just increasing the timeout limit.

Reply:
Well I still think that the timeout as suggested by this PR is the right choice. If you can improve the performance and lower the timeout along with that, then great. Otherwise it's always a clear improvement to bring timeouts in line with reality according to the rule of thumb that I mention

Reply:
Which failing spec does this timeout configuration even affect? The PR description doesn't give *any* context.

Reply:
Any of the HTTP server specs that use this method used to randomly fail.